### PR TITLE
wstunnel: 0.5.1.0 -> 9.6.2

### PR DESCRIPTION
**Pull request by NeverBehave**
## Description of changes

Add https://github.com/erebe/wstunnel , wstunnel rewritten in rust.

There is one exists in haskell but it seems the author abandon it: https://search.nixos.org/packages?channel=24.05&from=0&size=50&sort=relevance&type=packages&query=wstunnel

I renamed the package to `wstunnel-rust` for now, but maybe we should just replace it with the haskell one? @gebner

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by SuperSandro2000** [comment]
Please follow the [contributing guide](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#commit-conventions) when naming your commits.

File: pkgs/by-name/ws/wstunnel-rust/package.nix [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+  version = "9.6.1";
+
+  src = fetchFromGitHub {
+    owner = "erebe";
+    repo  = "wstunnel";
+    rev = "v${version}";
+    sha256 = "sha256-+tNQf53KUNIz9RFVTHqgXEKLxlXiQKKPZtXqD3C/t7I=";
+  };
+
+  cargoHash = "sha256-NER7NIWUXyWFUGcdZ5HLFaKaO9FNLPlG7jql08karS4=";
```
Comment (SuperSandro2000) [inline pkgs/by-name/ws/wstunnel-rust/package.nix:17]:
```suggestion
  cargoHash = "sha256-NER7NIWUXyWFUGcdZ5HLFaKaO9FNLPlG7jql08karS4=";

```

File: pkgs/by-name/ws/wstunnel-rust/package.nix [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+  meta = with lib; {
+    description = "Wstunnel uses the websocket protocol which is compatible with http in order to bypass firewalls and proxies. Wstunnel allows you to tunnel whatever traffic you want and access whatever resources/site you need.";
+    mainProgram = "wstunnel";
+    homepage    = "https://github.com/erebe/wstunnel";
+    license     = with licenses; [ bsd3Clear ];
+    maintainers = with maintainers; [ neverbehave ];
+    platforms   = platforms.linux;
+  };
+}
+
```
Comment (SuperSandro2000) [inline pkgs/by-name/ws/wstunnel-rust/package.nix:32]: there is one new line to much here

File: pkgs/by-name/ws/wstunnel-rust/package.nix [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+  };
+
+  cargoHash = "sha256-NER7NIWUXyWFUGcdZ5HLFaKaO9FNLPlG7jql08karS4=";
+  checkFlags = [
+    # make use of network connection
+    "--skip=tcp::tests::test_proxy_connection"
+  ];
+
+  meta = with lib; {
+    description = "Wstunnel uses the websocket protocol which is compatible with http in order to bypass firewalls and proxies. Wstunnel allows you to tunnel whatever traffic you want and access whatever resources/site you need.";
```
Comment (SuperSandro2000) [inline pkgs/by-name/ws/wstunnel-rust/package.nix:24]:
```suggestion
    description = "Wstunnel uses the websocket protocol in order to bypass firewalls and proxies";
```
we neeed to shorten that or move some things to longDescription.

**Comment by NeverBehave** [comment]
I will keep the name `wstunnel-rust` for now to avoid confusion with haskell package.

File: pkgs/by-name/ws/wstunnel-rust/package.nix [hunk 4]
```diff
@@ -0,0 +1,32 @@
+{ fetchFromGitHub
+, rustPlatform
+, lib
+}:
+
+rustPlatform.buildRustPackage rec {
+  pname = "wstunnel-rust";
```
Comment (NickCao) [inline pkgs/by-name/ws/wstunnel-rust/package.nix:7]: `pname` should be identical to upstream package name, and can be different from the attribute name, so `wstunnel` is fine.

File: pkgs/by-name/ws/wstunnel-rust/package.nix [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+  version = "9.6.1";
+
+  src = fetchFromGitHub {
+    owner = "erebe";
+    repo  = "wstunnel";
+    rev = "v${version}";
+    sha256 = "sha256-+tNQf53KUNIz9RFVTHqgXEKLxlXiQKKPZtXqD3C/t7I=";
```
Comment (NickCao) [inline pkgs/by-name/ws/wstunnel-rust/package.nix:14]:
```suggestion
    hash = "sha256-+tNQf53KUNIz9RFVTHqgXEKLxlXiQKKPZtXqD3C/t7I=";
```

File: pkgs/by-name/ws/wstunnel-rust/package.nix [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+  };
+
+  cargoHash = "sha256-NER7NIWUXyWFUGcdZ5HLFaKaO9FNLPlG7jql08karS4=";
+
+  checkFlags = [
+    # make use of network connection
+    "--skip=tcp::tests::test_proxy_connection"
+  ];
+
+  meta = with lib; {
```
Comment (NickCao) [inline pkgs/by-name/ws/wstunnel-rust/package.nix:24]: Avoid using `with lib;` if possible.

File: pkgs/by-name/ws/wstunnel-rust/package.nix [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+    description = "Wstunnel uses the websocket protocol which is compatible with http in order to bypass firewalls and proxies.";
```
Comment (NickCao) [inline pkgs/by-name/ws/wstunnel-rust/package.nix:25]: Description should not contain package name, should be short and concise, and shouldn't end with a period.

File: pkgs/by-name/ws/wstunnel-rust/package.nix [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+    mainProgram = "wstunnel";
+    homepage    = "https://github.com/erebe/wstunnel";
+    license     = with licenses; [ bsd3Clear ];
```
Comment (NickCao) [inline pkgs/by-name/ws/wstunnel-rust/package.nix:28]: Doesn't seem to be `bsd3Clear`, not sure which variant of bsd license either.

**Comment by h7x4** [comment]
> There is one exists in haskell but it seems the author abandon it: [search.nixos.org/packages?channel=24.05&from=0&size=50&sort=relevance&type=packages&query=wstunnel](https://search.nixos.org/packages?channel=24.05&from=0&size=50&sort=relevance&type=packages&query=wstunnel)
>
> I renamed the package to `wstunnel-rust` for now, but maybe we should just replace it with the haskell one?

> I will keep the name `wstunnel-rust` for now to avoid confusion with haskell package, but binary name intact as `wstunnel`.

According to the readme (https://github.com/erebe/wstunnel?tab=readme-ov-file#note), the project seems to have had a rewrite from haskell. This also checks out for hackage, with the author name being the same as the github project owner (https://hackage.haskell.org/package/wstunnel). Since this means that the haskell package is legacy code and this could be considered a "big update", I think the new package should just claim the `wstunnel` attribute, and then we drop the haskell package. There seems to be no haskell dependents, so there should not be any breakage.

File: pkgs/by-name/ws/wstunnel-rust/package.nix (hunk 8 above)
Comment (NeverBehave) [reply pkgs/by-name/ws/wstunnel-rust/package.nix:28]: How to include the license here? It does sounds like a BSD

Comment (NickCao) [reply pkgs/by-name/ws/wstunnel-rust/package.nix:28]: There is a concerning line: `All rights reserved`, better ask author for clarification.

Comment (NeverBehave) [reply pkgs/by-name/ws/wstunnel-rust/package.nix:28]: I think this is from https://spdx.org/licenses/BSD-3-Clause-Clear.html ?

Comment (NickCao) [reply pkgs/by-name/ws/wstunnel-rust/package.nix:28]: But missing the `NO EXPRESS OR IMPLIED LICENSES TO ANY PARTY'S PATENT RIGHTS ARE GRANTED BY THIS LICENSE` wording about patent.

Comment (NeverBehave) [reply pkgs/by-name/ws/wstunnel-rust/package.nix:28]: Open a discussion at the project: https://github.com/erebe/wstunnel/discussions/293

Comment (NeverBehave) [reply pkgs/by-name/ws/wstunnel-rust/package.nix]: mark as unfree until we have update from the author

Comment (NeverBehave) [reply pkgs/by-name/ws/wstunnel-rust/package.nix]: Author makes clear that it is BSD3

**Comment by h7x4** [comment]
There seems to be some cleanup needed for the `wstunnel` nixos module to function properly. Are you up to the task, or would you rather not touch it?

**Comment by NickCao** [comment]
Hash mismatch on ofborg:
```
error: hash mismatch in fixed-output derivation '/nix/store/7jxg6yq98d22qxvh4qsc0n9y0fbkqig4-wstunnel-9.6.1-vendor.tar.gz.drv':
         specified: sha256-NER7NIWUXyWFUGcdZ5HLFaKaO9FNLPlG7jql08karS4=
            got:    sha256-a0k1j3LemSXHVZeU5+UgcIjrK9ppAgF65ojnOPlU3iw=
```

**Comment by NeverBehave** [comment]
> There seems to be some cleanup needed for the `wstunnel` nixos module to function properly. Are you up to the task, or would you rather not touch it?

I think it only refer to the `pkgs.wstunnel` but CLI API should be the same. I didn't lookin to too much.
I am okay to work them thru but not in this PR if possible. @h7x4 how usually this should be handled.

**Comment by h7x4** [comment]
> I think it only refer to the pkgs.wstunnel but CLI API should be the same. I didn't lookin to too much.

I have just barely looked at the cli flags, and the first thing that hits me is that they have changed from pascalCase to kebab-case. The available flags also did not seem to match up with the old ones completely. The package bump will probably break the module completely.

> I am okay to work them thru but not in this PR if possible. @h7x4 how usually this should be handled?

Tbh, I'm not sure what would be the usual way to handle it, if such a way even exists. It's not often that an update involves rewriting the program from the bottom up in another language. But in order to not render the module completely useless on the master branch, I think we either should introduce the package under another attribute name (e.g. `wstunnel-rust`, like you initially proposed) and keep the old package until we are able to fix the module, or we just wait with merging the PR until the module has been fixed in this same PR.

**Comment by NeverBehave** [comment]
@h7x4 The original haskell package is actually marked broken and the module is not really usable. I could have the module package default pointed back the haskell one if that's okay, or I could include the module changes as well.

The reason I don't want to include that in this PR is 1) avoid bigger PR and testing headache 2) I am currently not familiar w/ testing nixos modules yet thus this will stale for a while until I put all pieces together..

**Comment by NeverBehave** [comment]
Actually I just found there is another PR around this: https://github.com/NixOS/nixpkgs/pull/289652

**Comment by NeverBehave** [comment]
The API changes a lot, I have to change the module a lot. E.g. structurally changed the host module definition. 
@alyaeanyx Would you open to that? Or you have better idea how this would proceed.

**Comment by NeverBehave** [comment]
I have tried building the module with some different configuration.nix w/ `nixos-rebuild build-vm --fast -I nixos-config=./configuration.nix -I nixpkgs=nixpkgs --show-trace` and fix some typo.

Since there isn't much feedback from the original maintainer, @NickCao Would you be able to take another quick look about the changes here? Since it involve some module changes, I am not sure if I should add some extra docs etc. 
Any advise would be appreciated.

File: nixos/modules/services/networking/wstunnel.nix [hunk 9]
```diff
@@ -66,10 +48,16 @@ let
       };
     };
 
-    verboseLogging = mkOption {
-      description = "Enable verbose logging.";
-      type = types.bool;
-      default = false;
+    loggingLevel = mkOption {
+      description = ''
+        passing to --log-lvl
```
Comment (NickCao) [inline nixos/modules/services/networking/wstunnel.nix:53]:
```suggestion
        Passed to --log-lvl
```

File: nixos/modules/services/networking/wstunnel.nix [hunk 10]
```diff
@@ -66,10 +48,16 @@ let
...
+
+        Control the log verbosity. i.e: TRACE, DEBUG, INFO, WARN, ERROR, OFF
+        for more details: https://docs.rs/tracing-subscriber/latest/tracing_subscriber/filter/struct.EnvFilter.html#example-syntax
```
Comment (NickCao) [inline nixos/modules/services/networking/wstunnel.nix:56]: We can use markdown formatting here for the url. e.g. `[EnvFilter](https://docs.rs/tracing-subscriber/latest/tracing_subscriber/filter/struct.EnvFilter.html#example-syntax)`

File: nixos/modules/services/networking/wstunnel.nix [hunk 11]
```diff
@@ -134,59 +123,36 @@ let
...
-        example = {
-          host = "example.com";
-        };
-      };
-
-      enableHTTPS = mkOption {
-        description = "Enable HTTPS when connecting to the server.";
-        type = types.bool;
-        default = true;
+        type = types.str;
```
Comment (NickCao) [inline nixos/modules/services/networking/wstunnel.nix:131]: Why not the `hostPortSubmodule`?

**Review by NickCao — COMMENTED** [review]
The package part looks good. But the inconsistent use of `str` and `localRemoteSubmodule` types in the module can use some rework.

File: nixos/modules/services/networking/wstunnel.nix (hunk 11 above)
Comment (NeverBehave) [reply nixos/modules/services/networking/wstunnel.nix:131]: Right now it has at least two protocol support: 

> <ws[s]|http[s]://wstunnel.server.com[:port]>
          Address of the wstunnel server
          You can either use websocket or http2 as transport protocol. Use websocket if you are unsure.
          Example: For websocket with TLS wss://wstunnel.example.com or without ws://wstunnel.example.com
                   For http2 with TLS https://wstunnel.example.com or without http://wstunnel.example.com

**Comment by NeverBehave** [comment]
> `localRemoteSubmodule`

This submodule is now replaced with str since the number of protocol/combination has expand a lot. Similar to `services.autossh.sessions.*.extraArguments`, where user specify the entire forward options instead of predefined submodule, seems to be a better approach imo for this situation.

**Review by NickCao — APPROVED** [review]
diff LGTM, please format the second commit as `nixos/wstunnel: ...`.

**Comment by NickCao** [comment]
And a few points for future improvements:
- The lack of `utils.escapeSystemdExecArgs` client unit.
- A nixos test.

**Pull request merged**